odxtools/encodestate.py

```diff
@@ -86,66 +87,141 @@ def emplace_atomic_value(
         is_highlow_byte_order: bool,
         used_mask: Optional[bytes],
     ) -> None:
         """Convert the internal_value to bytes and emplace this into the PDU"""
 
         raw_value: AtomicOdxType
 
-        # Check that bytes and strings actually fit into the bit length
+        # Deal with raw byte fields, ...
         if base_data_type == DataType.A_BYTEFIELD:
             if not isinstance(internal_value, (bytes, bytearray, SupportsBytes)):
-                odxraise()
-            if 8 * len(internal_value) > bit_length:
-                raise EncodeError(f"The bytefield {internal_value.hex()} is too large "
-                                  f"({len(internal_value)} bytes)."
-                                  f" The maximum length is {bit_length//8}.")
-            raw_value = bytes(internal_value)
-        elif base_data_type == DataType.A_ASCIISTRING:
-            if not isinstance(internal_value, str):
-                odxraise()
+                odxraise(f"{internal_value!r} is not a bytefield", EncodeError)
+                return
 
-            # The spec says ASCII, meaning only byte values 0-127.
-            # But in practice, vendors use iso-8859-1, aka latin-1
-            # reason being iso-8859-1 never fails since it has a valid
-            # character mapping for every possible byte sequence.
-            raw_value = internal_value.encode("iso-8859-1")
+            odxassert(base_type_encoding is None or base_type_encoding == Encoding.NONE)
 
-            if 8 * len(raw_value) > bit_length:
-                raise EncodeError(f"The string {repr(internal_value)} is too large."
-                                  f" The maximum number of characters is {bit_length//8}.")
-        elif base_data_type == DataType.A_UTF8STRING:
-            if not isinstance(internal_value, str):
-                odxraise()
+            raw_value = bytes(internal_value)
 
-            raw_value = internal_value.encode("utf-8")
+        # ... string types, ...
+        elif base_data_type in (DataType.A_UTF8STRING, DataType.A_ASCIISTRING,
+                                DataType.A_UNICODE2STRING):
+            if not isinstance(internal_value, str):
+                odxraise(f"The internal value {internal_value!r} is not a string", EncodeError)
+
+            if base_type_encoding == Encoding.UTF8:
+                raw_value = internal_value.encode("utf-8")
+            elif base_type_encoding == Encoding.UCS2:
+                text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                raw_value = internal_value.encode(text_encoding)
+            elif base_type_encoding == Encoding.ISO_8859_1:
+                raw_value = internal_value.encode("iso-8859-1")
+            elif base_type_encoding == Encoding.ISO_8859_2:
+                raw_value = internal_value.encode("iso-8859-2")
+            elif base_type_encoding == Encoding.WINDOWS_1252:
+                raw_value = internal_value.encode("cp1252")
```

Comment:
Copy-paste, extract logic into a function

Reply:
this is only _almost_ copy-and-paste: `DecodeState` needs to call `.decode()` instead of `.encode()`

Reply:
create function that takes the coded type as input, and return the correct string encoding name that python uses

Reply:
okay, I budge (also for the `MinMaxLengthType` issue below: https://github.com/mercedes-benz/odxtools/pull/381/commits/7de4a81490072c1e6a6287aef84a7e7a6d60477b . Thanks for the persistent nagging :)